modules/apps/document-library/document-library-test/src/testIntegration/java/com/liferay/document/library/change/tracking/test/DLFileVersionConstraintResolverTest.java

```diff
@@ -0,0 +1,160 @@
+			DLFolderConstants.DEFAULT_PARENT_FOLDER_ID,
+			RandomTestUtil.randomString(), RandomTestUtil.randomString(), false,
+			serviceContext);
+
+		byte[] bytes = TestDataConstants.TEST_BYTE_ARRAY;
+
+		InputStream inputStream = new ByteArrayInputStream(bytes);
+
+		DLFileEntry dlFileEntry = _dlFileEntryLocalService.addFileEntry(
+			null, _group.getCreatorUserId(), _group.getGroupId(),
+			_group.getGroupId(), dlFolder.getFolderId(),
+			RandomTestUtil.randomString(), ContentTypes.TEXT_PLAIN,
+			RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+			RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+			DLFileEntryTypeConstants.FILE_ENTRY_TYPE_ID_BASIC_DOCUMENT, null,
+			null, inputStream, bytes.length, null, null, null, serviceContext);
+
+		CTCollection ctCollection1 = _ctCollectionLocalService.addCTCollection(
+			null, TestPropsValues.getCompanyId(), TestPropsValues.getUserId(),
+			0, RandomTestUtil.randomString(), null);
+
+		try (SafeCloseable safeCloseable =
+				CTCollectionThreadLocal.setCTCollectionIdWithSafeCloseable(
+					ctCollection1.getCtCollectionId())) {
+
+			_dlFileEntryLocalService.updateFileEntry(
+				TestPropsValues.getUserId(), dlFileEntry.getFileEntryId(),
+				dlFileEntry.getFileName(), ContentTypes.TEXT_PLAIN,
+				RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+				RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+				DLVersionNumberIncrease.MAJOR,
+				DLFileEntryTypeConstants.FILE_ENTRY_TYPE_ID_BASIC_DOCUMENT,
+				null, null, inputStream, bytes.length, null, null, null,
+				serviceContext);
+		}
+
+		CTCollection ctCollection2 = _ctCollectionLocalService.addCTCollection(
+			null, TestPropsValues.getCompanyId(), TestPropsValues.getUserId(),
+			0, RandomTestUtil.randomString(), null);
+
+		try (SafeCloseable safeCloseable =
+				CTCollectionThreadLocal.setCTCollectionIdWithSafeCloseable(
+					ctCollection2.getCtCollectionId())) {
+
+			_dlFileEntryLocalService.updateFileEntry(
+				TestPropsValues.getUserId(), dlFileEntry.getFileEntryId(),
+				dlFileEntry.getFileName(), ContentTypes.TEXT_PLAIN,
+				RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+				RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+				DLVersionNumberIncrease.MAJOR,
+				DLFileEntryTypeConstants.FILE_ENTRY_TYPE_ID_BASIC_DOCUMENT,
+				null, null, inputStream, bytes.length, null, null, null,
+				serviceContext);
+		}
+
+		_ctCollectionService.publishCTCollection(
+			TestPropsValues.getUserId(), ctCollection1.getCtCollectionId());
+
+		_ctCollectionService.publishCTCollection(
+			TestPropsValues.getUserId(), ctCollection2.getCtCollectionId());
+
+		Map<Long, List<ConflictInfo>> conflictsMap1 =
+			_ctCollectionLocalService.checkConflicts(ctCollection1);
+
+		Assert.assertFalse(conflictsMap1.isEmpty());
+
+		Map<Long, List<ConflictInfo>> conflictsMap2 =
+			_ctCollectionLocalService.checkConflicts(ctCollection2);
+
+		Assert.assertFalse(conflictsMap2.isEmpty());
```

Comment:
Conflict checks should be done before publish.

Reply:
Do we still need to publish ctCollection1 then? Since the test will be checking the conflictsResolution before the publish and that's our goal, it is not needed, right?

Reply:
I was testing here and if the checks are done before publishing the ctcollection no conflicts are found, because the check looks at the conflicts [between the publication and the production](https://github.com/liferay-publications/liferay-portal/blob/b18182269962d953d40478dd906ad009f988a64c/modules/apps/change-tracking/change-tracking-service/src/main/java/com/liferay/change/tracking/service/impl/CTCollectionLocalServiceImpl.java#L172-L176). Should I put the check to ctCollection1 conflicts before publishing it and keep the check of ctcollection2 after publishing ctcollection1? This way it would check both situations

Reply:
Yes, you need to publish ctCollection1 in order to create the conflict. You won't be able to publish if there are conflicts; you'll get an error. So you'll do this for the test:

1. Publish ctCollection1, no conflict check is needed since conflict is not expected
2. Check conflicts for ctCollection2 and assert that conflicts exist. You don't need to publish ctCollection2 since we're expecting conflicts here.